cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/joiner/Join.java

```diff
@@ -0,0 +1,258 @@
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Join plugin to perform joins on structured records
+ */
+@Plugin(type = BatchJoiner.PLUGIN_TYPE)
+@Name("Join")
+public class Join extends BatchJoiner<StructuredRecord, StructuredRecord, StructuredRecord> {
+  public static final PluginClass PLUGIN_CLASS = getPluginClass();
+  private final Config config;
+  private Schema outputSchema;
+
+  public Join(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Map<String, Schema> inputSchemas = stageConfigurer.getInputSchemas();
+    stageConfigurer.setOutputSchema(getOutputSchema(inputSchemas));
+    config.validateConfig();
+  }
+
+  @Override
+  public StructuredRecord joinOn(String stageName, StructuredRecord record) throws Exception {
+    List<Schema.Field> fields = new ArrayList<>();
+    Schema schema = record.getSchema();
+
+    // TODO create output record based on fields properties
+    Map<String, List<String>> stageToJoinKey = config.getJoinKeys();
+    List<String> joinKeys = stageToJoinKey.get(stageName);
+    int i = 1;
+    for (String joinKey : joinKeys) {
+      Schema.Field joinField = Schema.Field.of("_" + i++, schema.getField(joinKey).getSchema());
+      fields.add(joinField);
+    }
+    Schema keySchema = Schema.recordOf("join.key", fields);
+    StructuredRecord.Builder keyRecordBuilder = StructuredRecord.builder(keySchema);
+    i = 1;
+    for (String joinKey : joinKeys) {
+      keyRecordBuilder.set("_" + i++, record.get(joinKey));
+    }
+
+    return keyRecordBuilder.build();
+  }
+
+  @Override
+  public JoinConfig getJoinConfig() {
+    return new JoinConfig(config.joinType, config.getNumOfInputs());
+  }
+
+  @Override
+  public void merge(StructuredRecord joinKey, List<JoinResult> joinResults, Emitter<StructuredRecord> emitter) {
+    Set<Schema.Field> fields = new HashSet<>();
+    StructuredRecord.Builder outRecordBuilder;
+
+    for (JoinResult joinResult : joinResults) {
+      if (outputSchema == null) {
+        outputSchema = getOutputSchema(fields, joinResult);
+      }
+      outRecordBuilder = StructuredRecord.builder(outputSchema);
+      for (JoinElement joinElement : joinResult.getJoinResult()) {
+        StructuredRecord record = (StructuredRecord) joinElement.getJoinValue();
+        for (Schema.Field field : record.getSchema().getFields()) {
+          outRecordBuilder.set(field.getName(), record.get(field.getName()));
+        }
+      }
+      emitter.emit(outRecordBuilder.build());
+    }
+  }
+
+  private Schema getOutputSchema(Set<Schema.Field> fields, JoinResult joinResult) {
```

Comment:
It will be good to not generate the output schema based on the join result. (I see that you are propagating output schema from configure time in the outer join branch)